presto-main/src/main/java/io/prestosql/sql/planner/ScopeAware.java

```diff
@@ -0,0 +1,156 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import io.prestosql.sql.analyzer.Analysis;
+import io.prestosql.sql.analyzer.ResolvedField;
+import io.prestosql.sql.analyzer.Scope;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.Node;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ScopeAware<T extends Node>
+{
+    private final Analysis analysis;
+    private final Scope scope;
+    private final T node;
+
+    private ScopeAware(Analysis analysis, Scope scope, T node)
+    {
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.scope = requireNonNull(scope, "scope is null");
+        this.node = requireNonNull(node, "node is null");
+    }
+
+    public static <T extends Node> ScopeAware<T> scopeAwareKey(T node, Analysis analysis, Scope scope)
+    {
+        return new ScopeAware<T>(analysis, scope, node);
+    }
+
+    public T getNode()
+    {
+        return node;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return hash(node);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ScopeAware<T> other = (ScopeAware<T>) o;
+        checkArgument(scope.isLocalScope(other.scope) || other.scope.isLocalScope(scope), "Expressions must be in the same local scope");
+
+        return equal(node, other.node);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ScopeAware(" + node + ")";
+    }
+
+    private boolean equal(Node left, Node right)
+    {
+        if (left instanceof Expression && right instanceof Expression) {
+            Expression leftExpression = (Expression) left;
+            Expression rightExpression = (Expression) right;
+            if (analysis.isColumnReference(leftExpression) && analysis.isColumnReference(rightExpression)) {
+                ResolvedField leftField = analysis.getResolvedField(leftExpression);
+                ResolvedField rightField = analysis.getResolvedField(rightExpression);
+
+                Scope leftScope = leftField.getScope();
+                Scope rightScope = rightField.getScope();
+
+                if (isLocalScope(leftScope) || isLocalScope(rightScope)) {
```

Comment:
Consider the following query:

```
SELECT -x AS x 
FROM (VALUES 1) t(x)
ORDER BY x
```

the `x` in the `SELECT` clause refers to `t.x`. The `x` in the `ORDER BY` clause refers to `AS x`. They have different field ids, but they can't be considered equal from syntax alone.

Reply:
Ok, so you probably could change the condition to
```
if (leftField.getFieldId().equals(rightField.getFieldId())) {
  return true;
}

if (!isLocalScope(leftScope) && !isLocalScope(rightScope)) {
  return false;
}

return leftExpression.equals(rightExpression);
```
it seems more natural and if also seems to be easier to fix `hash` below

Reply:
That's not equivalent. We need to be able to resolve references that *don't* have the same syntax to the same field. For instance:

```
SELECT (SELECT t.x), (SELECT x) FROM (VALUES 1) t(x)
```

Both subqueries are equivalent because `x` and `t.x` reference the same column from `t`, even though they have different syntax.